File: Semestrul 4/SDI/lab11/java/TriathlonServer/src/main/java/StartJsonServer.java
import server.ServiceImpl;
import triathlon.network.jsonprotocol.TriathlonClientJsonWorker;
import triathlon.persistence.impl.*;
import triathlon.services.ITriathlonServices;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Properties;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class StartJsonServer {
    private static final Logger logger = LogManager.getLogger(StartJsonServer.class);
    private static final int DEFAULT_PORT = 55555;
    private static final int THREAD_POOL_SIZE = 10;

    public static void main(String[] args) {
        var eventRepository = new EventRepositoryHibernateImpl();
        var participantrepository = new ParticipantRepositoryHibernateImpl();
        var resultrepository = new ResultRepositoryImpl(eventRepository, participantrepository);
        var refereerepository = new RefereeRepositoryImpl();
        var refereeDisciplineRepository = new RefereeDisciplineRepositoryImpl(refereerepository, eventRepository);

        ITriathlonServices serverImpl = new ServiceImpl(eventRepository, participantrepository, refereerepository, resultrepository, refereeDisciplineRepository);
        Properties serverProps = new Properties();
        try {
            serverProps.load(StartJsonServer.class.getResourceAsStream("/config.properties"));
            logger.info("Loaded client properties");
        } catch (IOException e) {
            logger.error("Cannot find triathlon-client.properties", e);
        }
        int port = serverProps.getProperty("triathlon.server.port") != null ? Integer.parseInt(serverProps.getProperty("triathlon.server.port")) : DEFAULT_PORT;
        try {
            if (args.length > 0) {
                port = Integer.parseInt(args[0]);
            }
        } catch (NumberFormatException e) {
            logger.error("Invalid port number", e);
            System.out.println("Invalid port number. Using default port " + DEFAULT_PORT);
        }

        logger.info("Starting server on port: " + port);
        System.out.println("Starting server on port: " + port);

        ExecutorService executorService = Executors.newFixedThreadPool(THREAD_POOL_SIZE);

        try (ServerSocket serverSocket = new ServerSocket(port)) {
            logger.info("Server started. Waiting for clients...");
            System.out.println("Server started. Waiting for clients...");

            while (true) {
                try {
                    Socket clientSocket = serverSocket.accept();
                    logger.info("Client connected from: " + clientSocket.getInetAddress());
                    System.out.println("Client connected from: " + clientSocket.getInetAddress());

                    TriathlonClientJsonWorker worker = new TriathlonClientJsonWorker(serverImpl, clientSocket);
                    executorService.submit(worker);
                } catch (IOException e) {
                    logger.error("Error accepting client", e);
                    System.out.println("Error accepting client: " + e.getMessage());
                }
            }
        } catch (IOException e) {
            logger.error("Error starting server", e);
            System.out.println("Error starting server: " + e.getMessage());
        } finally {
            executorService.shutdown();
            logger.info("Server stopped");
            System.out.println("Server stopped");
        }
    }
}

File: Semestrul 4/SDI/lab09-10/server .net/TriathlonServer/StartJsonServer.cs
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Configuration;
using log4net;
using Triathlon.Server;
using Triathlon.Network.JsonProtocol;
using Triathlon.Persistence.Impl;
using Triathlon.Services;

namespace Triathlon.Server;

public class StartJsonServer
{
    private static readonly ILog _logger = LogManager.GetLogger(typeof(StartJsonServer));
    private const int DEFAULT_PORT = 55555;
    private const int THREAD_POOL_SIZE = 10;

    public static void Main(string[] args)
    {
        log4net.Config.XmlConfigurator.Configure(new FileInfo("log4net.config"));
        var eventRepository = new EventRepositoryImpl();
        var participantRepository = new ParticipantRepositoryImpl();
        var resultRepository = new ResultRepositoryImpl(eventRepository, participantRepository);
        var refereeRepository = new RefereeRepositoryImpl();
        var refereeDisciplineRepository = new RefereeDisciplineRepositoryImpl(refereeRepository, eventRepository);

        ITriathlonServices serverImpl = new ServiceImplementation(eventRepository, participantRepository, refereeRepository, resultRepository, refereeDisciplineRepository);

                 int port = DEFAULT_PORT;
        try
        {
            string portString = ConfigurationManager.AppSettings["triathlon.server.port"];
            if (!string.IsNullOrEmpty(portString))
            {
                port = int.Parse(portString);
                _logger.Info("Using port from configuration: " + port);
            }
        }
        catch (Exception ex)
        {
            _logger.Error("Error reading configuration", ex);
        }

                 try
        {
            if (args.Length > 0)
            {
                port = int.Parse(args[0]);
            }
        }
        catch (FormatException ex)
        {
            _logger.Error("Invalid port number", ex);
            Console.WriteLine("Invalid port number. Using default port " + DEFAULT_PORT);
        }

        _logger.Info("Starting server on port: " + port);
        Console.WriteLine("Starting server on port: " + port);

                 ThreadPool.SetMaxThreads(THREAD_POOL_SIZE, THREAD_POOL_SIZE);

                 using (var cancellationTokenSource = new CancellationTokenSource())
        {
            try
            {
                                 TcpListener listener = new TcpListener(IPAddress.Any, port);
                listener.Start();

                _logger.Info("Server started. Waiting for clients...");
                Console.WriteLine("Server started. Waiting for clients...");

                                 Console.CancelKeyPress += (sender, e) => {
                    e.Cancel = true;
                    cancellationTokenSource.Cancel();
                };

                                 while (!cancellationTokenSource.Token.IsCancellationRequested)
                {
                    try
                    {
                        TcpClient client = listener.AcceptTcpClient();
                        _logger.Info($"Client connected from: {((IPEndPoint)client.Client.RemoteEndPoint).Address}");
                        Console.WriteLine($"Client connected from: {((IPEndPoint)client.Client.RemoteEndPoint).Address}");

                                                 TriathlonClientJsonWorker worker = new TriathlonClientJsonWorker(serverImpl, client);
                        ThreadPool.QueueUserWorkItem(_ => {
                            try
                            {
                                worker.Run();
                            }
                            catch (Exception ex)
                            {
                                _logger.Error("Error in worker thread", ex);
                            }
                        });
                    }
                    catch (SocketException ex)
                    {
                        if (!cancellationTokenSource.Token.IsCancellationRequested)
                        {
                            _logger.Error("Error accepting client", ex);
                            Console.WriteLine("Error accepting client: " + ex.Message);
                        }
                    }
                    catch (Exception ex)
                    {
                        _logger.Error("Unexpected error accepting client", ex);
                        Console.WriteLine("Unexpected error accepting client: " + ex.Message);
                    }
                }

                                 listener.Stop();
            }
            catch (Exception ex)
            {
                _logger.Error("Error starting server", ex);
                Console.WriteLine("Error starting server: " + ex.Message);
            }
            finally
            {
                _logger.Info("Server stopped");
                Console.WriteLine("Server stopped");
            }
        }
    }
}